feat(core): Add composite foreign key support for associations

gtamasi opened the pull request:
<!--
Thanks for wanting to fix something on Sequelize - we already love you!
Please fill in the template below.
If unsure about something, just do as best as you're able.
-->

## Pull Request Checklist

<!-- Please make sure to review and check all of these items: -->

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? <!-- Put PR link here -->
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

<!-- NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open. -->

## Description of Changes

Currently, Sequelize primarily supports single-column foreign keys. However, many database schemas utilize composite foreign keys to enforce data integrity and model complex relationships more accurately. This feature addresses this limitation and provides greater flexibility in defining database mappings.

Issue: #311 

**Changes Introduced:**

* `foreignKey` option in `belongsTo`, `hasOne`, and `hasMany` associations contains an extra field called `keys` which holds the keys used for composite keys. This allows specifying multiple columns that constitute the foreign key in the source model.  
* The keys can be defined by either with a string which is attribute of the model or by adding `sourceKey` and `targetKey` attribute. The logic is the same as if you use the `targetKey` attribute in case of single column key.
* Updated association logic to correctly handle composite foreign key constraints** during data retrieval, creation, update, and deletion operations.


With this design, the aim was the backwards compatibility so this PR contains no breaking changes. 


**Example Usage:**

```typescript
// Define the target model with a composite primary key
   const User = this.sequelize.define('User', {
              userId: {
                type: DataTypes.INTEGER,
                primaryKey: true,
              },
              tenantId: {
                type: DataTypes.INTEGER,
                primaryKey: true,
              },
              username: DataTypes.STRING,
            });
            const Address = this.sequelize.define(
              'Address',
              {
                addressId: {
                  type: DataTypes.INTEGER,
                  primaryKey: true,
                  autoIncrement: true,
                },
                zipCode: DataTypes.STRING,
              },
            );
   Address.belongsTo(User, { foreignKey: { keys: ['userId', 'tenantId'] } });



// Define Belongs-to-many

      this.User = this.sequelize.define('User', {
        userId: {
          type: DataTypes.INTEGER,
          primaryKey: true,
          autoIncrement: true,
        },
        tenantId: {
          type: DataTypes.INTEGER,
          primaryKey: true,
          autoIncrement: false,
        },
        name: DataTypes.STRING,
      });
      this.Project = this.sequelize.define('Project', {
        projectId: {
          type: DataTypes.INTEGER,
          primaryKey: true,
          autoIncrement: true,
        },
        tenantId: {
          type: DataTypes.INTEGER,
          primaryKey: true,
          autoIncrement: false,
        },
        name: DataTypes.STRING,
      });

      this.User.belongsToMany(this.Project, {
        through: 'user_projects',
        as: 'Projects',
        inverse: {
          as: 'Users',
        },
        foreignKey: {
          keys: ['userId', 'tenantId'],
        },
        otherKey: {
          keys: ['projectId', 'tenantId'],
        },
      });
    });


```


## Notes
In SQLite3, you only get autoincrement behavior when only one integer column is the primary key. Therefore, you cannot define a composite primary key when one of the column defined as `AUTOINCREMENT`. 
You can get a similar result by defining a column as the only primary key, but then adding an additional unique constraint on columns.


## List of Breaking Changes



<!-- If you have caused any breaking changes, you should list them below. -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added support for composite (multi-column) foreign keys across all association types.
  * Enabled NULLS FIRST/LAST ordering options for index fields.
  * Added SQL logging to development environment for debugging.

* **Bug Fixes**
  * Fixed foreign key constraint handling during database schema synchronization.
  * Improved attribute handling in associated model includes.
  * Fixed composite key mapping in eager-loaded associations.

* **Type Updates**
  * Refined return types for model aggregation methods (max, min, sum) for improved type safety.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wd-alejandroescobedogarcia committed:
add composite key findByPk support via object

wd-alejandroescobedogarcia committed:
prettier changes

wd-alejandroescobedogarcia committed:
code review changes

wd-alejandroescobedogarcia committed:
Merge branch 'main' into findbypk-composite-support

wd-alejandroescobedogarcia committed:
prettier changes

wd-alejandroescobedogarcia committed:
Merge branch 'main' into findbypk-composite-support

wd-alejandroescobedogarcia committed:
Merge branch 'main' into findbypk-composite-support

wd-alejandroescobedogarcia committed:
check if model has pk first

wd-alejandroescobedogarcia committed:
change misused property

papandreou committed:
Change the package names to sequelize-{postgres,core}-papandreou

(cherry picked from commit 6eb101e610be3ab6ede00fe0222a6724d0d2b262)
(cherry picked from commit c48086b7b56b6e3ff9b6a6246415a3c43abfb21a)
(cherry picked from commit 00fad6a0bfa325d5d426948729a9e183166a6233)

wd-alejandroescobedogarcia committed:
consolidate changes from other branches

wd-alejandroescobedogarcia committed:
update types

wd-alejandroescobedogarcia committed:
add constraintName property

wd-alejandroescobedogarcia committed:
cleanup models file

wd-alejandroescobedogarcia committed:
add missing tests

papandreou committed:
Replace workspace:* urls

(cherry picked from commit df19318b06d4058db6b2d1a944daaea3b557fa63)

wd-alejandroescobedogarcia committed:
use ? to check if it is safe to push to array in hasmany

wd-alejandroescobedogarcia committed:
Fix issue with includes and hasmany

papandreou committed:
Merge commit '0717a42ad' into sequelize-core-papandreou9
This is the "known good" state of the findbypk-composite-support branch
* commit '0717a42ad':
  change misused property
  check if model has pk first
  prettier changes
  code review changes
  prettier changes
  add composite key findByPk support via object

papandreou committed:
Merge branch 'feature/composite-primary-key-support-and-associations-patch' into sequelize-core-papandreou9

* feature/composite-primary-key-support-and-associations-patch:
  Fix issue with includes and hasmany
  use ? to check if it is safe to push to array in hasmany
  add missing tests
  cleanup models file
  add constraintName property
  update types
  consolidate changes from other branches

papandreou committed:
Releases sequelize-{core,postgres}-papandreou@7.0.0-alpha.44-patch1

papandreou committed:
Merge branch 'main' into sequelize-core-papandreou9

* main:
  fix(core): Adjust model validator types yet again (#17689)

papandreou committed:
Releases sequelize-{core,postgres}-papandreou@7.0.0-alpha.44-patch2

gtamasi committed:
Merge remote-tracking branch 'origin/sequelize-core-papandreou9' into gtamasi/sequelize-composite-foreign-key-support

gtamasi committed:
remove custom package names

gtamasi committed:
refactor: update foreign key handling in associations to support composite keys

gtamasi committed:
fix: update dependency to use workspace versioning

gtamasi committed:
fix: downgrade pg dependency to version 8.11.3

gtamasi committed:
fix: prevent simultaneous use of foreignKey.name and foreignKey.keys in association options

gtamasi committed:
fix: throw AssociationError for simultaneous use of foreignKey.name and foreignKey.keys

ephys commented:
Hi! I'd love to see this feature implemented, but before going too far in the implementation, make sure to take a look at the design decisions that were discussed near the end of https://github.com/sequelize/sequelize/issues/311 :)

gtamasi commented:
Hey @ephys! Would you mind helping us out with giving some early feedback about what we have already done so far. As far I can see we don't deviate significantly from the design you have discussed in the issue. It would be nice to know what else is missing to get this change accepted. 

What is missed that I know:
- Belongs-to-many implementation
- Documentation

gtamasi committed:
fix: helper func

gtamasi committed:
fix: add support for composite foreign keys in BelongsToMany association

gtamasi committed:
Revert "fix: add support for composite foreign keys in BelongsToMany association"

This reverts commit 33f52c165919c14640bfdc102e7af8545bc17c39.

gtamasi committed:
Revert "refactor: update foreign key handling in associations to support composite keys"

This reverts commit bddc9c6da33b8734228f1c8d8ea9208e64bcde94.

gtamasi committed:
add new test

gtamasi committed:
wip

gtamasi committed:
fix reference

gtamasi committed:
add new getters

gtamasi committed:
refactor: enhance composite key handling in query generator

gtamasi committed:
fix: use association.source

gtamasi committed:
fix: update identifierField and add foreignIdentifierFields getter in BelongsToManyAssociation

gtamasi committed:
refactor: streamline composite key handling in query generator

gtamasi committed:
revert

gtamasi committed:
don't skip

gtamasi committed:
fix: optimize where clause construction in BelongsToManyAssociation

gtamasi committed:
fix: consolidate where clause construction in BelongsToManyAssociation

gtamasi committed:
fix:typo

gtamasi committed:
feat: add Key interface and update foreign key types in associations

gtamasi committed:
feat: enhance foreign key handling with support for multiple foreign keys and composite foreign key options

gtamasi committed:
docs: add detailed comments for foreign key attributes in BelongsToMany and BelongsTo associations

gtamasi committed:
Merge branch 'main' into gtamasi/sequelize-composite-foreign-key-support

gtamasi committed:
fix: fix issue caused by conflict resolve

Signed-off-by: gtamasi <gergo.tamasi@workday.com>

gtamasi committed:
lint: remove unneccesary lines

gtamasi committed:
lint: prettier fix

gtamasi committed:
fix: test

gtamasi committed:
fix: add exception when dialect is sqlite3

gtamasi commented on packages/core/src/associations/base.ts:
```diff
@@ -280,17 +287,30 @@ export interface ForeignKeyOptions<ForeignKey extends string>
+  /**
+   * The pairs of the foreign key attributes used for composite foreign keys.
+   */
+  keys?: ForeignKey[] | Array<CompositeForeignKeysOptions<ForeignKey, SourceKey>>;
```
We consciously went down this road to make this change backwards-compatible. We discussed the API design choice proposed in #311, but realized that by having `foreignKeys` as an array would make it impossible to define constraints and cascading for the composite key itself.

gtamasi committed:
refactor: change interface to type for Key definition

gtamasi committed:
lint

gtamasi committed:
lint

gtamasi committed:
refactor: update ForeignKeyOptions interface to simplify type definitions

gtamasi committed:
lint

gtamasi committed:
fix: db2 onUpdate constraint

gtamasi committed:
fix

gtamasi committed:
Merge upstream/main into gtamasi/sequelize-composite-foreign-key-support

Resolved conflict in packages/core/src/abstract-dialect/query-generator.js:
wrapped upstream's improved alias resolution logic (isMinified branch,
non-root-parent branch) in the composite-key `each(attrSources, ...)` loop,
replacing singular `identSource` with `identSources[i]`.

gtamasi committed:
fix: add oracle expectation to composite foreign key generateJoin test

gtamasi committed:
fix: composite FK regressions after upstream merge

- Avoid calling singular `association.sourceKeyField` inside the composite
  `each(attrSources)` loop in `generateThroughJoin`. For BelongsToMany the
  singular getter resolves a single targetKey that does not exist on the
  target model and crashes in `getColumnName`. Use the indexed plural
  `sourceKeyFields[i]` for the composite case instead.

- Fix `referencedColumnNames` ordering in postgres `showConstraintsQuery`.
  The cross-join with `information_schema.constraint_column_usage` plus
  upstream's `ORDER BY ccu.column_name` causes referenced columns to come
  back alphabetically rather than in FK definition order, breaking the
  composite FK sync tests. Resolve the referenced column per source
  position via `pg_constraint.confkey[kcu.ordinal_position]` so each row
  reports the correctly paired column.

gtamasi committed:
Revert "fix: composite FK regressions after upstream merge"

This reverts commit 6a30bb49667953646029ce027ac70bc5cceef073.

gtamasi committed:
fix: avoid singular sourceKeyField getter crash on composite through joins

The `isMinified` branch of `generateThroughJoin` called
`association.sourceKeyField` (the singular getter), which throws for
composite `BelongsToMany` associations because no single `sourceKey` is
defined. Resolve the column via the indexed plural getter
(`sourceKeyFields?.[i]`) for composite associations and the singular
getter only for non-composite ones.

gtamasi committed:
Delete .nvmrc

gtamasi committed:
Delete dev/postgres/latest/docker-compose.yml

gtamasi committed:
fix(postgres): preserve definition order of composite FK referencedColumnNames

`information_schema.constraint_column_usage` (ccu) is a flat list with
no positional column. Joining it to `key_column_usage` (kcu) on the
constraint name produces an N×M cross-product for composite FKs, and
upstream's `ORDER BY ccu.column_name` then surfaces referenced columns
in alphabetical rather than definition order. The Set-based JS
accumulator in `showConstraints` preserves first-seen order, so the
final array is alphabetical (e.g. `['tenantId', 'userId']` instead of
`['userId', 'tenantId']`).

Resolve the referenced column for each kcu row directly via
`pg_constraint.confkey[kcu.ordinal_position]`, which is the positional
attnum array indexed identically to `conkey`/`kcu.ordinal_position`.
The cross-product still exists, but every duplicate row at the same
ordinal_position now shares the same referencedColumnNames value, so
de-duplication preserves definition order.

gtamasi committed:
Revert "Delete dev/postgres/latest/docker-compose.yml"

This reverts commit 407439f6267fc3c24acc8e9ebf875fd6b0aeb017.

gtamasi committed:
Revert "fix(postgres): preserve definition order of composite FK referencedColumnNames"

This reverts commit af60302de2cb9a656a28e9ee835bf4de7c03412e.

gtamasi committed:
fix(postgres): re-order composite FK referencedColumnNames via pg_constraint

`information_schema.constraint_column_usage` (ccu) is a flat list with
no positional column. After kcu × ccu cross-joins for composite FKs and
upstream's `ORDER BY ccu.column_name`, the JS Set accumulator in
`showConstraints` surfaces `referencedColumnNames` in alphabetical
order rather than definition order (e.g. `['tenantId', 'userId']`
instead of `['userId', 'tenantId']`).

Override `showConstraints` in the postgres query interface to detect
composite FKs in the standard result and re-fetch the position-ordered
referenced columns directly from `pg_constraint.confkey`. Joining by
OID through `pg_class`/`pg_namespace` sidesteps the broken
`pgc.conrelid::regclass::text` comparison in `showConstraintsQuery`,
which fails for non-public schemas and quoted identifiers (the cause
of the regressions in the prior attempt).

The follow-up query runs only when at least one composite FK is
present, so single-key FK callers pay no extra round trip.

wd-alejandroescobedogarcia committed:
fix some hasOne issues from code rabbit

wd-alejandroescobedogarcia committed:
handle composite foreign keys in HasOne#set
